cluster/channel.go

```diff
@@ -0,0 +1,149 @@
+		Name:        "alertmanager_oversized_gossip_message_failure_total",
+		Help:        "Number of oversized gossip message sends that failed.",
+		ConstLabels: prometheus.Labels{"key": key},
+	})
+	oversizeGossipMessageSentTotal := prometheus.NewCounter(prometheus.CounterOpts{
+		Name:        "alertmanager_oversized_gossip_message_sent_total",
+		Help:        "Number of oversized gossip message sent.",
+		ConstLabels: prometheus.Labels{"key": key},
+	})
+	oversizeGossipMessageDroppedTotal := prometheus.NewCounter(prometheus.CounterOpts{
+		Name:        "alertmanager_oversized_gossip_message_dropped_total",
+		Help:        "Number of oversized gossip messages that were dropped due to a full message queue.",
+		ConstLabels: prometheus.Labels{"key": key},
+	})
+	oversizeGossipDuration := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name:        "alertmanager_oversize_gossip_message_duration_seconds",
+		Help:        "Duration of oversized gossip message requests.",
+		ConstLabels: prometheus.Labels{"key": key},
+	})
+
+	reg.MustRegister(oversizeGossipDuration, oversizeGossipMessageFailureTotal, oversizeGossipMessageDroppedTotal, oversizeGossipMessageSentTotal)
+
+	c := &Channel{
+		key:                               key,
+		send:                              send,
+		peers:                             peers,
+		logger:                            logger,
+		msgc:                              make(chan []byte, 200),
+		sendOversize:                      sendOversize,
+		oversizeGossipMessageFailureTotal: oversizeGossipMessageFailureTotal,
+		oversizeGossipMessageDroppedTotal: oversizeGossipMessageDroppedTotal,
+		oversizeGossipMessageSentTotal:    oversizeGossipMessageSentTotal,
+		oversizeGossipDuration:            oversizeGossipDuration,
+	}
+
+	go c.handleOverSizedMessages(stopc)
+
+	return c
+}
+
+// handleOverSizedMessages prevents memberlist from opening too many parallel
+// TCP connections to its peers.
+func (c *Channel) handleOverSizedMessages(stopc chan struct{}) {
+	var wg sync.WaitGroup
+	for {
+		select {
+		case b := <-c.msgc:
+			for _, n := range c.peers() {
+				wg.Add(1)
+				go func(n *memberlist.Node) {
+					defer wg.Done()
+					c.oversizeGossipMessageSentTotal.Inc()
+					start := time.Now()
+					if err := c.sendOversize(n, b); err != nil {
+						level.Debug(c.logger).Log("msg", "failed to send reliable", "key", c.key, "node", n, "err", err)
+						c.oversizeGossipMessageFailureTotal.Inc()
+						return
+					}
+					c.oversizeGossipDuration.Observe(time.Since(start).Seconds())
+				}(n)
+			}
+
+			wg.Wait()
+		case <-stopc:
+			return
+		}
+	}
+}
+
+// Broadcast enqueues a message for broadcasting.
+func (c *Channel) Broadcast(b []byte) {
+	b, err := proto.Marshal(&clusterpb.Part{Key: c.key, Data: b})
+	if err != nil {
+		return
+	}
+
+	if OversizedMessage(b) {
+		select {
+		case c.msgc <- b:
+		default:
+			level.Warn(c.logger).Log("msg", "oversized gossip channel full")
```

Comment:
With the dropped metric (which is better than log lines anyway), I think this should be a debug now to not spam the log.